src/app/services/todos-store.service.ts

```diff
@@ -103,32 +103,54 @@ export class TodosStoreService {
     const todo = this.todos.find((t) => t.id === id);
     this.todos = this.todos.filter((t) => t.id !== id);
 
-    // TODO: Undo operation
-    // Timeout server remove by notifier duration
-    // And then reset using notifier's action
-
-    if (serverRemove) {
-      this.todosService.remove(id).subscribe(
-        () => {},
-        (error) => {
-          this.notifier.open(`Error: ${error.message}`);
-          setTimeout(() => (this.todos = [...this.todos, todo]));
-        }
-      );
-    }
+    // Prompt user if undo is necessary
+    const mode = { undo: false };
+    const msDuration = 3000;
+    const revertCb = () => {
+      mode.undo = true;
+      this.todos = [...this.todos, todo];
+    };
+    this.notifier.undoAlert(mode, revertCb, `Removed Todo.`, msDuration);
+
+    // If not undoing, send server request after alert duration
+    setTimeout(() => {
+      if (serverRemove && !mode.undo) {
+        this.todosService.remove(id).subscribe(
+          () => {},
```

Comment:
if `undoAlert` is moved here, then the wrapping `setTimeout` can be deleted and `msDuration` moved to the `notification.service`